// @gera2ld/format-json/dist/index.common.js
/*! @gera2ld/format-json v0.2.1 | MIT License */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e['default'] : e; }

var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);

exports.ItemTypes = void 0;

(function (ItemTypes) {
  ItemTypes["MULTILINE"] = "MULTILINE";
  ItemTypes["SINGLELINE"] = "SINGLELINE";
  ItemTypes["KEY"] = "KEY";
  ItemTypes["COMMA"] = "COMMA";
  ItemTypes["BR"] = "BR";
})(exports.ItemTypes || (exports.ItemTypes = {}));

const COMMA = {
  type: exports.ItemTypes.COMMA,
  value: ','
};
const BR = {
  type: exports.ItemTypes.BR,
  value: '\n'
};
const charMapBase = {
  '\\': '\\\\',
  '\r': '\\r',
  '\t': '\\t'
};

const charMapQuote = _extends__default({}, charMapBase, {
  '\'': '\\\'',
  '\n': '\\n'
});

const charMapTemplate = _extends__default({}, charMapBase, {
  '`': '\\`'
});

function quoteString(str, {
  quote,
  quoteAsNeeded,
  template
}) {
  if (template && /\n/.test(str)) {
    const quoted = str.replace(/[\\`\r\t]/g, m => charMapTemplate[m]);
    return `\`${quoted}\``;
  }

  if (!quoteAsNeeded || /\W/.test(str)) {
    const re = new RegExp(`[\\\\\\r\\n\\t${quote}]`, 'g');
    const quoted = str.replace(re, m => charMapQuote[m]);
    return quote + quoted + quote;
  }

  return str;
}

function getSpace(level, indent) {
  return {
    type: 'space',
    value: ' '.repeat(indent * level)
  };
}

function renderArray(data, options, path = []) {
  const level = path.length;
  const arr = [];
  const ret = {
    type: exports.ItemTypes.MULTILINE,
    separator: [COMMA],
    data: arr,
    path
  };
  arr.push({
    value: '['
  });

  if (data.length) {
    const rendered = data.map((item, i) => render(item, options, [...path, i]));
    arr.push(...(options.indent ? [BR] : []), getSpace(level + 1, options.indent), ...join(rendered, options, level + 1), ...(options.indent ? [BR] : []), getSpace(level, options.indent));
  } else {
    ret.type = exports.ItemTypes.SINGLELINE;
  }

  arr.push({
    value: ']'
  });
  return ret;
}

function renderObject(data, options, path = []) {
  const level = path.length;
  const arr = [];
  const ret = {
    type: exports.ItemTypes.MULTILINE,
    separator: [COMMA],
    data: arr,
    path
  };
  arr.push({
    value: '{'
  });
  const rendered = Object.keys(data).map(key => {
    const subpath = [...path, key];
    const keyItem = {
      type: exports.ItemTypes.KEY,
      data: [{
        value: quoteString(key, options),
        type: 'key'
      }],
      separator: [{
        value: ':'
      }],
      path: subpath
    };
    options.onData == null ? void 0 : options.onData(keyItem);
    return [keyItem, render(data[key], options, subpath)];
  }).reduce((res, cur) => [...res, ...cur], []);

  if (rendered.length) {
    arr.push(...(options.indent ? [BR] : []), getSpace(level + 1, options.indent), ...join(rendered, options, level + 1), ...(options.indent ? [BR] : []), getSpace(level, options.indent));
  } else {
    ret.type = exports.ItemTypes.SINGLELINE;
  }

  arr.push({
    value: '}'
  });
  return ret;
}

function render(data, options, path = []) {
  let result;

  if (Array.isArray(data)) {
    result = renderArray(data, options, path);
  } else if (data === null) {
    result = {
      type: exports.ItemTypes.SINGLELINE,
      separator: [COMMA],
      data: [{
        value: data,
        type: 'null'
      }],
      path
    };
  } else if (typeof data === 'object') {
    result = renderObject(data, options, path);
  } else {
    result = {
      type: exports.ItemTypes.SINGLELINE,
      separator: [COMMA],
      data: [{
        value: typeof data === 'string' ? quoteString(data, _extends__default({}, options, {
          quoteAsNeeded: false
        })) : data
      }],
      path
    };
  }

  options.onData == null ? void 0 : options.onData(result);
  return result;
}

function join(rendered, options, level) {
  const arr = [];

  for (let i = 0; i < rendered.length; i += 1) {
    const item = rendered[i];
    const next = rendered[i + 1];
    if (item.data) arr.push(...item.data); // trailing separators

    if (item.separator && (next || options.trailing)) {
      arr.push(...item.separator);
    }

    if (next) {
      if (item.type === exports.ItemTypes.KEY) {
        if (options.indent) arr.push({
          value: ' '
        });
      } else {
        arr.push(...(options.indent ? [BR] : []), getSpace(level, options.indent));
      }
    }
  }

  return arr;
}

function format(data, options) {
  const renderOptions = _extends__default({
    indent: 0,
    quoteAsNeeded: false,
    quote: '"',
    trailing: false,
    template: false
  }, options);

  const rendered = render(data, renderOptions);
  return (rendered.data || []).map(({
    value
  }) => `${value}`).join('');
}

exports.format = format;
exports.render = render;
